Simplenote/NotesListController.swift

```diff
@@ -0,0 +1,213 @@
+import Foundation
+import CoreData
+import SimplenoteFoundation
+import SimplenoteSearch
+
+
+// MARK: - NotesListController
+//
+class NotesListController: NSObject {
+
+    /// Core Data Kung Fu
+    ///
+    private let viewContext: NSManagedObjectContext
+    private lazy var notesController = ResultsController<Note>(viewContext: viewContext,
+                                                               matching: state.predicateForNotes(filter: filter),
+                                                               sortedBy: state.descriptorsForNotes(sortMode: sortMode))
+
+    /// FSM: Active State
+    ///
+    private(set) var state: NotesListState = .results {
+        didSet {
+            guard oldValue != state else {
+                return
+            }
+
+            refreshEverything()
+        }
+    }
+
+    /// Filter: Applied when we're not in Search Mode
+    ///
+    var filter: NotesListFilter = .everything {
+        didSet {
+            guard oldValue != filter else {
+                return
+            }
+
+            refreshPredicates()
+        }
+    }
+
+    /// Refreshes the Internal State, so that Search Results matching the specified Keyword will be filtered out.
+    /// - Important: It's up to the caller to invoke `performFetch`!!
+    ///
+    var searchKeyword: String? {
+        get {
+            guard case let .searching(keyword) = self.state else {
+                return nil
+            }
+
+            return keyword
+        }
+        set {
+            guard let keyword = newValue, !keyword.isEmpty else {
+                state = .results
+                return
+            }
+
+            state = .searching(keyword: keyword)
+        }
+    }
+
+    /// SortMode: Results Mode
+    ///
+    var sortMode: SortMode = .alphabeticallyAscending {
+        didSet {
+            guard oldValue != sortMode else {
+                return
+            }
+
+            refreshSortDescriptors()
+        }
+    }
+
+    /// SortMode: Search Mode
+    ///
+    var searchSortMode: SortMode = .alphabeticallyAscending {
+        didSet {
+            guard case .searching = state, oldValue != searchSortMode else {
+                return
+            }
+
+            refreshSortDescriptors()
+        }
+    }
```

Comment:
Both Sort Modes will map to different User Defaults properties (this will be wired by the Sort Bar).

It's up to the ListController which one should be applied (otherwise we'd need to check on the state of the ListController before assigning the property).